services/src/main/java/org/keycloak/keys/JavaKeystoreKeyProvider.java

```diff
@@ -43,46 +48,57 @@
 import java.util.Optional;
 import java.util.Set;
 import java.util.stream.Collectors;
+import java.util.stream.Stream;
 
 /**
  * @author <a href="mailto:sthorger@redhat.com">Stian Thorgersen</a>
  */
-public class JavaKeystoreKeyProvider extends AbstractRsaKeyProvider {
+public class JavaKeystoreKeyProvider implements KeyProvider {
+
+    private final KeyStatus status;
+
+    private final ComponentModel model;
+
+    private final KeyWrapper key;
+
+    private final String algorithm;
 
     public JavaKeystoreKeyProvider(RealmModel realm, ComponentModel model) {
-        super(realm, model);
+        this.model = model;
+        this.status = KeyStatus.from(model.get(Attributes.ACTIVE_KEY, true), model.get(Attributes.ENABLED_KEY, true));
+
+        String defaultAlgorithmKey = KeyUse.ENC.name().equals(model.get(Attributes.KEY_USE)) ? JWEConstants.RSA_OAEP : Algorithm.RS256;
+        this.algorithm = model.get(Attributes.ALGORITHM_KEY, defaultAlgorithmKey);
+
+        if (model.hasNote(KeyWrapper.class.getName())) {
+            key = model.getNote(KeyWrapper.class.getName());
+        } else {
+            key = loadKey(realm, model);
+            model.setNote(KeyWrapper.class.getName(), key);
+        }
     }
 
-    @Override
     protected KeyWrapper loadKey(RealmModel realm, ComponentModel model) {
         String keystorePath = model.get(JavaKeystoreKeyProviderFactory.KEYSTORE_KEY);
         try (FileInputStream is = new FileInputStream(keystorePath)) {
-            // Use "JKS" as default type for backwards compatibility
-            String keystoreType = KeystoreUtil.getKeystoreType(model.get(JavaKeystoreKeyProviderFactory.KEYSTORE_TYPE_KEY), keystorePath, "JKS");
-            KeyStore keyStore = KeyStore.getInstance(keystoreType);
-            keyStore.load(is, model.get(JavaKeystoreKeyProviderFactory.KEYSTORE_PASSWORD_KEY).toCharArray());
-
+            KeyStore keyStore = loadKeyStore(is, keystorePath);
             String keyAlias = model.get(JavaKeystoreKeyProviderFactory.KEY_ALIAS_KEY);
-            PrivateKey privateKey = (PrivateKey) keyStore.getKey(keyAlias, model.get(JavaKeystoreKeyProviderFactory.KEY_PASSWORD_KEY).toCharArray());
-            PublicKey publicKey = KeyUtils.extractPublicKey(privateKey);
 
-            KeyPair keyPair = new KeyPair(publicKey, privateKey);
-
-            X509Certificate certificate = (X509Certificate) keyStore.getCertificate(keyAlias);
-            if (certificate == null) {
-                certificate = CertificateUtils.generateV1SelfSignedCertificate(keyPair, realm.getName());
-            }
-
-            KeyUse keyUse = KeyUse.valueOf(model.get(Attributes.KEY_USE, KeyUse.SIG.getSpecName()).toUpperCase());
-
-            return createKeyWrapper(keyPair, certificate, loadCertificateChain(keyStore, keyAlias), keyUse);
+            return switch (algorithm) {
+                case Algorithm.HS256, Algorithm.HS384, Algorithm.HS512, Algorithm.RS256, Algorithm.RS384, Algorithm.RS512 ->
```

Comment:
I suppose the `HS*` algorithms here should be replaced by `PS*` algorithms on this line? As I see that `Attributes.RS_ALGORITHM_PROPERTY` supports keys of type `Algorithm.RS256, Algorithm.RS384, Algorithm.RS512, Algorithm.PS256, Algorithm.PS384, Algorithm.PS512` but it does not support `Algorithm.HS256, Algorithm.HS384, Algorithm.HS51` .

Reply:
You're right, I fixes it. see https://github.com/keycloak/keycloak/pull/27030/commits/19e7795fc24caec6780afc0d5beaed8c0257a3d4